### PR TITLE
fix conversion of empty `Dates.CompoundPeriod()` to zero units

### LINKED ISSUE #50193: Method error when converting empty `Dates.CompoundPeriod` period
MWE:
```
julia> using Dates
julia> convert(Nanosecond, Dates.CompoundPeriod())
ERROR: MethodError: reducing over an empty collection is not allowed; consider supplying `init` to the reducer
Stacktrace:
  [1] reduce_empty(op::Base.MappingRF{Type{Nanosecond}, typeof(Base.add_sum)}, #unused#::Type{Period})
    @ Base ./reduce.jl:356
  [2] reduce_empty_iter
    @ ./reduce.jl:379 [inlined]
  [3] mapreduce_empty_iter(f::Type, op::Function, itr::Vector{Period}, ItrEltype::Base.HasEltype)
    @ Base ./reduce.jl:375
  [4] _mapreduce(f::Type{Nanosecond}, op::typeof(Base.add_sum), #unused#::IndexLinear, A::Vector{Period})
    @ Base ./reduce.jl:427
  [5] _mapreduce_dim
    @ ./reducedim.jl:365 [inlined]
  [6] #mapreduce#765
    @ ./reducedim.jl:357 [inlined]
  [7] mapreduce
    @ ./reducedim.jl:357 [inlined]
  [8] #_sum#775
    @ ./reducedim.jl:999 [inlined]
  [9] _sum
    @ ./reducedim.jl:999 [inlined]
 [10] #sum#773
    @ ./reducedim.jl:995 [inlined]
 [11] sum
    @ ./reducedim.jl:995 [inlined]
 [12] convert(#unused#::Type{Nanosecond}, x::Dates.CompoundPeriod)
    @ Dates ~/.julia/juliaup/julia-1.8.5+0.aarch64.apple.darwin14/share/julia/stdlib/v1.8/Dates/src/periods.jl:363
 [13] top-level scope
    @ REPL[6]:1
```

Installed via juliaup:
```julia
julia> versioninfo()
Julia Version 1.8.5
Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.5.0)
  CPU: 8 × Apple M1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 1 on 4 virtual cores
```

### PR DESCRIPTION
**Pull request by vmpyr**
Closes: #50193 

WRT the fix proposed by @Seelengrab in his comment. Thanks to him for providing the whole fix. This PR fixes the issue and adds a test for the new functionality.

### CODE REVIEW

**Review by stevengj — APPROVED** [review]

**Comment by vmpyr** [comment]
I think this can be merged now.

File: stdlib/Dates/src/periods.jl [hunk 1]
```diff
@@ -325,7 +325,7 @@ end
 Base.show(io::IO,x::CompoundPeriod) = print(io, string(x))
 
 Base.convert(::Type{T}, x::CompoundPeriod) where T<:Period =
-    isconcretetype(T) ? sum(T, x.periods) : throw(MethodError(convert,(T,x)))
+    isconcretetype(T) ? sum(T, x.periods; init = zero(T)) : throw(MethodError(convert,(T,x)))
```
Comment (stevengj) [inline stdlib/Dates/src/periods.jl:328]: It seems like `sum(T::Type, iterator)` should use `zero(T)` automatically for empty collections.   Wouldn't that be a better fix?

Comment (stevengj) [reply stdlib/Dates/src/periods.jl:328]: In particular, we could define
```jl
Base.mapreduce_empty(::Type{T}, ::typeof(Base.add_sum), ::Any) where {T} = zero(T)
Base.reduce_empty_iter(::Base.MappingRF{Type{T},typeof(Base.add_sum)}, ::Base.EltypeUnknown) where {T} = zero(T)
```

Comment (stevengj) [reply stdlib/Dates/src/periods.jl:328]: ref #36262

Comment (stevengj) [reply stdlib/Dates/src/periods.jl:328]: Oh, this option was discussed and rejected in #20561.

**Comment by ararslan** [comment]
I'd argue that this is a bugfix; any chance it can be backported for 1.10.5, @KristofferC?

**Pull request merged**